lib/segment/src/vector_storage/mmap_type.rs

```diff
@@ -0,0 +1,504 @@
+        self.mmap.flusher()
+    }
+}
+
+impl Deref for MmapBitSlice {
+    type Target = BitSlice;
+
+    fn deref(&self) -> &BitSlice {
+        &self.mmap
+    }
+}
+
+impl DerefMut for MmapBitSlice {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.mmap
+    }
+}
+
+/// Get a second mutable reference for type `T` from the given mmap
+///
+/// # Warning
+///
+/// The returned reference is unbounded. The user must ensure it never outlives the `mmap` type.
+///
+/// # Safety
+///
+/// - unsafe because we create a second (unbounded) mutable reference
+/// - malformed data in the mmap may break the transmuted type `T` resulting in undefined behavior
+///
+/// # Panics
+///
+/// - panics when the size of the mmap doesn't match size `T`
+/// - panics when the mmap data is not correctly aligned for type `T`
+unsafe fn mmap_to_type_unbounded<'unbnd, T>(mmap: &mut MmapMut) -> &'unbnd mut T
+where
+    T: Sized,
+{
+    // Obtain unbounded bytes slice into mmap
+    let bytes: &'unbnd mut [u8] = {
+        let slice = mmap.deref_mut();
+        slice::from_raw_parts_mut(slice.as_mut_ptr(), slice.len())
+    };
+
+    // Assert alignment and size
+    assert_alignment::<_, T>(bytes);
+    debug_assert_eq!(mmap.len(), bytes.len());
+    assert_eq!(bytes.len(), mem::size_of::<T>());
+
+    let ptr = bytes.as_ptr() as *mut T;
+    unsafe { &mut *ptr }
+}
+
+/// Get a second mutable reference for a slice of type `T` from the given mmap
+///
+/// A (non-zero) header size in bytes may be provided to omit from the BitSlice data.
+///
+/// # Warning
+///
+/// The returned reference is unbounded. The user must ensure it never outlives the `mmap` type.
+///
+/// # Safety
+///
+/// - unsafe because we create a second (unbounded) mutable reference
+/// - malformed data in the mmap may break the transmuted slice for type `T` resulting in undefined
+///   behavior
+///
+/// # Panics
+///
+/// - panics when the size of the mmap isn't a multiple of size `T`
+/// - panics when the mmap data is not correctly aligned for type `T`
+/// - panics when the header size isn't a multiple of size `T`
+unsafe fn mmap_to_slice_unbounded<'unbnd, T>(
+    mmap: &mut MmapMut,
+    header_size: usize,
+) -> &'unbnd mut [T]
+where
+    T: Sized,
+{
+    // Obtain unbounded bytes slice into mmap
+    let bytes: &'unbnd mut [u8] = {
+        let slice = mmap.deref_mut();
+        &mut slice::from_raw_parts_mut(slice.as_mut_ptr(), slice.len())[header_size..]
+    };
+
+    // Assert alignment and size
+    assert_alignment::<_, T>(bytes);
+    debug_assert_eq!(mmap.len(), bytes.len() + header_size);
+    debug_assert_eq!(
+        header_size % mem::size_of::<T>(),
+        0,
+        "header not multiple of size T",
+    );
+    assert_eq!(
+        bytes.len() % mem::size_of::<T>(),
+        0,
+        "bytes not multiple of size T",
+    );
+
+    // Transmute slice types
+    slice::from_raw_parts_mut(bytes.as_ptr() as *mut T, bytes.len() / mem::size_of::<T>())
```

Comment:
This is UB under stacked borrows: &[u8] -> *const u8 -> *mut T -> &mut[T].
Using `bytes.as_mut_ptr()` instead of `bytes.as_ptr()` should be sound though.

Equivalent UB in `mmap_to_type_unbounded`, also fixed by using `as_mut_ptr` instead of `as_ptr`

Reply:
Good catch! Thanks a lot.

Reply:
Just wanted to mention I found this by changing MmapMut to Vec and running the tests with miri. 
Just a cool trick that may be useful in the future. Also, the rest of the logic was sound according to miri, so that might be a nice assurance.

Reply:
:pray: